src/symbolize/gimli.rs

```diff
@@ -295,17 +310,31 @@ struct LibrarySegment {
     len: usize,
 }
 
-#[cfg(target_os = "aix")]
 fn create_mapping(lib: &Library) -> Option<Mapping> {
-    let name = &lib.name;
-    let member_name = &lib.member_name;
-    Mapping::new(name.as_ref(), member_name)
+    cfg_if::cfg_if! {
+        if #[cfg(target_os = "aix")] {
+            Mapping::new(lib.name.as_ref(), &lib.member_name)
+        } else if #[cfg(target_os = "android")] {
+            Mapping::new_android(lib.name.as_ref(), lib.zip_offset)
+        } else {
+            Mapping::new(lib.name.as_ref())
+        }
+    }
 }
 
-#[cfg(not(target_os = "aix"))]
-fn create_mapping(lib: &Library) -> Option<Mapping> {
-    let name = &lib.name;
-    Mapping::new(name.as_ref())
+/// Try to extract the archive path from an "embedded" library path
+/// (e.g. `/path/to/my.apk` from `/path/to/my.apk!/mylib.so`).
+///
+/// Returns `None` if the path does not contain a `!/` separator.
+#[cfg(target_os = "android")]
+fn extract_zip_path_android(path: &mystd::ffi::OsStr) -> Option<&mystd::ffi::OsStr> {
+    use mystd::os::unix::ffi::OsStrExt;
+
+    path.as_bytes()
+        .windows(2)
+        .enumerate()
+        .find(|(_, chunk)| chunk == b"!/")
```

Comment:
So we had a discussion about how the path can include multiple instances of `!/` due to both having an APK embedded .so and also just a directory with an `~/excited!` name.

...Because of this, it could even have... three.
However, if I understand the path structure and APK structure correctly, the `!/` that splits the embedded library must always be the *last* one... right?

...right?

...uh, if I'm right (I hope I am...) then we should probably use `.rev()` before we `.find()`?

Reply:
Actually I think it uses the first one: https://cs.android.com/android/platform/superproject/main/+/daa7e5ca6b9d2ae543e60be31502c302d08304e1:bionic/linker/linker.cpp;l=932

```cpp
  // Treat an '!/' separator inside a path as the separator between the name
  // of the zip file on disk and the subdirectory to search within it.
  // For example, if path is "foo.zip!/bar/bas/x.so", then we search for
  // "bar/bas/x.so" within "foo.zip".
  const char* const separator = strstr(path, kZipFileSeparator);
  if (separator == nullptr) {
    return -1;
  }
```

Reply:
*hmm*

Reply:
tempted to file the most contrived bug report

Reply:
I suppose we should match their bug with our own for the moment, but I'm still really curious if a `.so` can involve the `!/` sequence *inside* the APK zip?

Reply:
Seems to work fine, at least using the linker's CLI:

```
emu64xa:/data/local/tmp $ unzip -l android-test.apk                                                                                                                                                                                            
Archive:  android-test.apk
  Length      Date    Time    Name
---------  ---------- -----   ----
    52305  2024-09-08 02:07   suffix/3
    54461  2024-09-08 02:07   suffix/4
    52728  2024-09-08 02:07   prefix/1
    51323  2024-09-08 02:07   prefix/5
    21392  2024-09-08 02:07   prefix/4
    64481  2024-09-08 02:07   prefix/0
    11233  2024-09-08 02:07   suffix/8
      754  2024-09-08 02:07   suffix/7
    12335  2024-09-08 02:07   prefix/7
    50870  2024-09-08 02:07   prefix/6
      835  2024-09-08 02:07   suffix/2
    23008  2024-09-08 02:07   suffix/0
    21131  2024-09-08 02:07   suffix/9
 10560600  2024-09-08 02:05   somedir!/anotherone!/mylib.so
    11260  2024-09-08 02:07   prefix/3
    17046  2024-09-08 02:07   suffix/1
    43742  2024-09-08 02:07   prefix/8
    20599  2024-09-08 02:07   suffix/5
    54497  2024-09-08 02:07   prefix/2
     1370  2024-09-08 02:07   suffix/6
    26675  2024-09-08 02:07   prefix/9
---------                     -------
 11152645                     21 files
emu64xa:/data/local/tmp $ linker64 /data/local/tmp/android-test.apk!/somedir!/anotherone!/mylib.so
frame #0  - 0x00000073e9ee8d83bb - backtrace::backtrace::libunwind::trace::h89414344d67efe6b
                                 @ /build/src/backtrace/libunwind.rs:116
                                 - backtrace::backtrace::trace_unsynchronized::h78287cff64b72793
                                 @ /build/src/backtrace/mod.rs:66
[... clipped backtrace output ...]
frame #12 - 0x00000073e9ee8db1fe - main
frame #13 - 0x00000073e9ed3099f0 - __libc_init
```

Reply:
welp, there goes my hypothesis.